File: android/src/main/java/cn/qiuxiang/react/amap3d/AMapMarker.java
package cn.qiuxiang.react.amap3d;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.view.View;

import com.amap.api.maps.AMap;
import com.amap.api.maps.model.BitmapDescriptor;
import com.amap.api.maps.model.BitmapDescriptorFactory;
import com.amap.api.maps.model.LatLng;
import com.amap.api.maps.model.Marker;
import com.amap.api.maps.model.MarkerOptions;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.facebook.react.views.view.ReactViewGroup;

import java.util.HashMap;
import java.util.Map;

@SuppressLint("ViewConstructor")
public class AMapMarker extends ReactViewGroup {
    private static final Map<String, Float> COLORS;

    static {
        COLORS = new HashMap<>();
        COLORS.put("HUE_AZURE", BitmapDescriptorFactory.HUE_AZURE);
        COLORS.put("HUE_BLUE", BitmapDescriptorFactory.HUE_BLUE);
        COLORS.put("HUE_CYAN", BitmapDescriptorFactory.HUE_CYAN);
        COLORS.put("HUE_GREEN", BitmapDescriptorFactory.HUE_GREEN);
        COLORS.put("HUE_MAGENTA", BitmapDescriptorFactory.HUE_MAGENTA);
        COLORS.put("HUE_ORANGE", BitmapDescriptorFactory.HUE_ORANGE);
        COLORS.put("HUE_RED", BitmapDescriptorFactory.HUE_RED);
        COLORS.put("HUE_ROSE", BitmapDescriptorFactory.HUE_ROSE);
        COLORS.put("HUE_VIOLET", BitmapDescriptorFactory.HUE_VIOLET);
        COLORS.put("HUE_YELLOW", BitmapDescriptorFactory.HUE_YELLOW);
    }

    private ReactViewGroup infoWindow;
    private Marker marker;
    private LatLng position;
    private String title = "";
    private String snippet = "";
    private boolean flat;
    private float opacity;
    private boolean draggable;
    private boolean selected;
    private boolean infoWindowEnabled;
    private BitmapDescriptor bitmapDescriptor;
    private RCTEventEmitter eventEmitter;

    public AMapMarker(ThemedReactContext context) {
        super(context);
        eventEmitter = context.getJSModule(RCTEventEmitter.class);
    }

    public void addToMap(AMap map) {
        marker = map.addMarker(getMarkerOptions());
        if (selected) {
            marker.showInfoWindow();
        } else {
            marker.hideInfoWindow();
        }
    }

    public String getMarkerId() {
        return marker.getId();
    }

    private MarkerOptions getMarkerOptions() {
        return new MarkerOptions()
                .setFlat(flat)
                .icon(bitmapDescriptor)
                .alpha(opacity)
                .draggable(draggable)
                .position(position)
                .title(title)
                .infoWindowEnable(infoWindowEnabled)
                .snippet(snippet);
    }

    public void setTitle(String title) {
        this.title = title;
        if (marker != null) {
            marker.setTitle(title);
        }
    }

    public void setSnippet(String snippet) {
        this.snippet = snippet;
        if (marker != null) {
            marker.setSnippet(snippet);
        }
    }

    public void setCoordinate(ReadableMap coordinate) {
        position = new LatLng(coordinate.getDouble("latitude"), coordinate.getDouble("longitude"));
        if (marker != null) {
            marker.setPosition(position);
        }
    }

    public void setFlat(boolean flat) {
        this.flat = flat;
        if (marker != null) {
            marker.setFlat(flat);
        }
    }

    public void setOpacity(float opacity) {
        this.opacity = opacity;
        if (marker != null) {
            marker.setAlpha(opacity);
        }
    }

    public void setDraggable(boolean draggable) {
        this.draggable = draggable;
        if (marker != null) {
            marker.setDraggable(draggable);
        }
    }

    public void setIcon(String icon) {
        if (icon.startsWith("HUE_")) {
            bitmapDescriptor = BitmapDescriptorFactory.defaultMarker(COLORS.get(icon));
            if (marker != null) {
                marker.setIcon(bitmapDescriptor);
            }
        }
    }

    public void sendEvent(String name, WritableMap data) {
        eventEmitter.receiveEvent(getId(), name, data);
    }

    public void setSelected(boolean selected) {
        this.selected = selected;
        if (marker != null) {
            if (selected) {
                marker.showInfoWindow();
            } else {
                marker.hideInfoWindow();
            }
        }
    }

    public ReactViewGroup getInfoWindow() {
        return infoWindow;
    }

    public void setInfoWindow(ReactViewGroup view) {
        infoWindow = view;
    }

    public void setIconView(final AMapOverlay overlay) {
        overlay.addOnLayoutChangeListener(new OnLayoutChangeListener() {
            @Override
            public void onLayoutChange(View view, int i, int i1, int i2, int i3, int i4, int i5, int i6, int i7) {
                updateIcon(overlay);
            }
        });
        overlay.setOnUpdateListener(new AMapOverlay.OnUpdateListener() {
            @Override
            public void onUpdate() {
                updateIcon(overlay);
            }
        });
    }

    private void updateIcon(AMapOverlay overlay) {
        Bitmap bitmap = Bitmap.createBitmap(
                overlay.getWidth(),
                overlay.getHeight(),
                Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        overlay.draw(canvas);
        bitmapDescriptor = BitmapDescriptorFactory.fromBitmap(bitmap);
        if (marker != null) {
            marker.setIcon(bitmapDescriptor);
        }
    }

    public void setEnabledInfoWindow(boolean enabled) {
        infoWindowEnabled = enabled;
    }
}

File: android/src/main/java/cn/qiuxiang/react/amap3d/AMapMarker.kt
package cn.qiuxiang.react.amap3d

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import com.amap.api.maps.AMap
import com.amap.api.maps.model.*
import com.facebook.react.views.view.ReactViewGroup

class AMapMarker(context: Context) : ReactViewGroup(context) {
    companion object {
        private val COLORS = mapOf(
                "AZURE" to BitmapDescriptorFactory.HUE_AZURE,
                "BLUE" to BitmapDescriptorFactory.HUE_BLUE,
                "CYAN" to BitmapDescriptorFactory.HUE_CYAN,
                "GREEN" to BitmapDescriptorFactory.HUE_GREEN,
                "MAGENTA" to BitmapDescriptorFactory.HUE_MAGENTA,
                "ORANGE" to BitmapDescriptorFactory.HUE_ORANGE,
                "RED" to BitmapDescriptorFactory.HUE_RED,
                "ROSE" to BitmapDescriptorFactory.HUE_ROSE,
                "VIOLET" to BitmapDescriptorFactory.HUE_VIOLET,
                "YELLOW" to BitmapDescriptorFactory.HUE_YELLOW
        )
    }

    var infoWindow: AMapOverlay? = null

    var infoWindowEnabled: Boolean = true
        set(value) {
            field = value
            marker?.isInfoWindowEnable = value
        }

    var marker: Marker? = null
        private set

    var position: LatLng? = null
        set(value) {
            field = value
            marker?.position = value
        }

    var zIndex: Float = 0.0f
        set(value) {
            field = value
            marker?.zIndex = value
        }

    var title = ""
        set(value) {
            field = value
            marker?.title = value
        }

    var snippet = ""
        set(value) {
            field = value
            marker?.snippet = value
        }

    var flat: Boolean = false
        set(value) {
            field = value
            marker?.isFlat = value
        }

    var opacity: Float = 1f
        set(value) {
            field = value
            marker?.alpha = value
        }

    var draggable: Boolean = false
        set(value) {
            field = value
            marker?.isDraggable = value
        }

    var clickable_: Boolean = true
        set(value) {
            field = value
            marker?.isClickable = value
        }

    var active: Boolean = false
        set(value) {
            field = value
            if (value) {
                marker?.showInfoWindow()
            } else {
                marker?.hideInfoWindow()
            }
        }

    private var bitmapDescriptor: BitmapDescriptor? = null

    fun addToMap(map: AMap) {
        marker = map.addMarker(MarkerOptions()
                .setFlat(flat)
                .icon(bitmapDescriptor)
                .alpha(opacity)
                .draggable(draggable)
                .position(position)
                .title(title)
                .infoWindowEnable(infoWindowEnabled)
                .snippet(snippet)
                .zIndex(zIndex))

        if (active) {
            marker?.showInfoWindow()
        } else {
            marker?.hideInfoWindow()
        }

        marker?.isClickable = this.clickable_
    }

    fun setIcon(icon: String) {
        bitmapDescriptor = COLORS[icon.toUpperCase()]?.let {
            BitmapDescriptorFactory.defaultMarker(it)
        }
        marker?.setIcon(bitmapDescriptor)
    }

    fun setIconView(overlay: AMapOverlay) {
        overlay.addOnLayoutChangeListener { _, _, _, _, _, _, _, _, _ -> updateIconView(overlay) }
        overlay.onUpdate {
            updateIconView(overlay)
        }
    }

    /**
     * TODO: 如果 IconView 里包含 Image，由于不知道 Image 什么时候加载完毕，会导致 Image 可能无法正确显示
     */
    private fun updateIconView(overlay: AMapOverlay) {
        val bitmap = Bitmap.createBitmap(
                overlay.width, overlay.height, Bitmap.Config.ARGB_8888)
        overlay.draw(Canvas(bitmap))
        bitmapDescriptor = BitmapDescriptorFactory.fromBitmap(bitmap)
        marker?.setIcon(bitmapDescriptor)
    }
}
